[docs] Add a tutorial for the pass manager

zhiics opened the pull request:
#3480 

This PR adds a tutorial for the recently merged pass manager infra.

cc @tqchen @jroesch @weberlo @slyubomirsky @wweic @yongwww @yzhliu @icemelon9 @MarisaKirisame

anijain2305 commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+
+Relay features a series of optimization passes which improve performance metrics
+of models such as mean inference, memory footprint, power consumption, or for
+specific devices. There are a suite of standard optimizations as well as machine
```
"or for specific devices" seems odd to read

anijain2305 commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+have custom passes correctly scheduled without having to modify a single
+handcrafted pass order.
+
+One the other hand, modern deep learning frameworks, such as Pytorch and MXNet
```
One -> On

anijain2305 commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+    class PassContextNode : public RelayNode {
+     public:
+      ErrorReporter err_reporter;
+      int opt_level{2};
```
Why is opt_level here? Isn't defining it in PassInfoNode sufficient?

zhiics replied:
yeah, each pass is associated with an opt_level. So you will know if a certain pass will be enabled when users provide a opt_level in the configuration context. This is different from what we usually have in a compiler.

anijain2305 commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+
+The pass manager is designed in a hierarchical manner, and it could work at
+various granularity of a Relay program. A pure virtual class ``PassNode`` is
+introduced to sever as the base of the different optimization passes. This class
```
sever -> serve

anijain2305 commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+Module-level Passes
+^^^^^^^^^^^^^^^^^^^
+
+Module level passes are geared mainly for global and inter-procedural
```
What is the example of IPO in this context?

zhiics replied:
we have inlining could be viewed as one although we don't really have multiple translation units.

anijain2305 commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+
+Function-level passes are used to implement various intra function level
+optimizations for a given Relay module. It fetches one function at a time from
+the function list of a module for optimization and yields a rewritten Relay
```
If there is a link for relay module description, it will be good to add here.

anijain2305 commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+indetermined that whether this pass is enabled, its ``opt_level`` will be
+checked. This pass will be enabled and therefore executed only when its
+optimization level not less than the configured optimization level in the pass
+context.
```
So, opt_level will be ignored, if user just specifies that I want to run Pass A, correct? Just like gcc -opt_A works even if we dont provide -O3 etc.

zhiics replied:
yes

shoubhik commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+==================================
+
+Relay features a series of optimization passes which improve performance metrics
+of models such as mean inference, memory footprint, power consumption, or for
```
or power consumption for specific devices.

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+
+#) providing a user friendly way to debug optimization passes.
+
+#) alleviating developers's from manually and respectively resolving the
```
```suggestion
#) alleviating developers from manually and respectively resolving the
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+Each subclass itself could act as a pass manager. For instance, they could glob
+the require passes and execute them or build a dependency graph based on the
+given meta data. The full definition of them could be found in
+``src/relay/pass/pass_manager.cc``
```
Could directly make this a link to the [source file](https://github.com/dmlc/tvm/blob/master/src/relay/pass/pass_manager.cc) using one of [these methods](http://docutils.sourceforge.net/docs/user/rst/quickref.html#hyperlink-targets)

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+passes for execution. It also has a disabled list that contains the list of
+disabled passes. For example, we may have a configuration which performs all
+passes at ``opt_level=3`` with some disabled passes using disabled_pass=xx
+provided by ``PassContext``. Now using ``SequentailPass``, we could glob all
```
```suggestion
provided by ``PassContext``. Now using ``SequentialPass``, we could glob all
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+SequentialPass is similar to Pytorch nn.Sequential that contains a host of
+passes for execution. It also has a disabled list that contains the list of
+disabled passes. For example, we may have a configuration which performs all
+passes at ``opt_level=3`` with some disabled passes using disabled_pass=xx
```
```suggestion
passes at ``opt_level=3`` with some disabled passes using ``disabled_pass=xx``
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+      Module operator()(const Module& mod, const PassContext& pass_ctx) const final;
+    };
+
+Only a few passes currently in Relay is put in this group. For example,
```
```suggestion
Only a few passes currently in Relay are put in this group. For example,
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+``FoldScaleAxis`` requires to dispatch ``ForwardFoldScaleAxis`` and
+``BackwardFoldScaleAxis`` internally. In addition, ``BackwardFoldScaleAxis`` is
+recommended to be fulfilled first. This pass, hence, is an ideal candidate for
+``SequentailPass``.
```
```suggestion
``SequentialPass``.
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+The following code shows how individual passes in a sequential pass is invoked.
+Essentially, we sequentially execute each pass in a sequential pass using the
+order that they were appended to the pass list without the consideration of
+their orders. The pass phase ordering problem might be considered in the future.
```
"using the order that they were appended to the pass list without the consideration of their orders" sounds like a contradiction. maybe trim off the end of the sentence?

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+The following code shows how individual passes in a sequential pass is invoked.
+Essentially, we sequentially execute each pass in a sequential pass using the
+order that they were appended to the pass list without the consideration of
+their orders. The pass phase ordering problem might be considered in the future.
```
Is it relevant to bring up the pass phase ordering problem here?

zhiics replied:
Removed

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+Upon the invocation of a pass, we first check if this pass is enabled. This is
+done by first checking if the pass is explicitly disabled by a user, followed by
+inspecting if it is specified as a required pass by the user. If it is still
+indetermined that whether this pass is enabled, its ``opt_level`` will be
```
```suggestion
undetermined whether this pass is enabled, its ``opt_level`` will be
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+inspecting if it is specified as a required pass by the user. If it is still
+indetermined that whether this pass is enabled, its ``opt_level`` will be
+checked. This pass will be enabled and therefore executed only when its
+optimization level not less than the configured optimization level in the pass
```
```suggestion
optimization level is not less than the configured optimization level in the pass
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+First, we create a simple Relay program, ``y = f(x)``. Then, we build a module
+based on the function. After creating the module, we instantiate a sequential
+pass object which contains some standard Relay optimization passes, including
+typer inference, dead code elimination, common subexpression elimination, and
```
```suggestion
type inference, dead code elimination, common subexpression elimination, and
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+``CreateFunctionPass``. The ``pass_func`` is returned as a packed function that
+invokes the ``Expr`` to ``Expr`` API on each function in a Relay module. ``{}``
+indicates that no prerequisite is required for this pass. Otherwise, the pass
+develop has to identify and list them.
```
```suggestion
developer has to identify and list them.
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+    }  // namespace transform
+
+To allow other C++ modules to apply this pass, we declare a free function in
+`include/tvm/relay/transform.h` as the following:
```
```suggestion
``include/tvm/relay/transform.h`` as the following:
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+       def __call__(self, mod):
+           return _transform.RunPass(self, mod)
+
+Some auxiliary APIs are provided to enable to easy creation of passes from
```
```suggestion
Some auxiliary APIs are provided to enable easy creation of passes from
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+           return _transform.RunPass(self, mod)
+
+Some auxiliary APIs are provided to enable to easy creation of passes from
+the Python frontend and let the pass manager to control the execution. For
```
```suggestion
the Python frontend and to let the pass manager control the execution. For
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+example, ``module_pass``, ``function_pass``, and ``sequential`` are provided to
+users so that they can customize their own pass or pass pipeline.
+
+For all the passes that implemented in the C++ backend, we provide
```
```suggestion
For all the passes that are implemented in the C++ backend, we provide
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+    mod = relay.Module()
+    mod = module_pass(mod)
+
+Correspondingly, we also offer such functionality for `function_pass`. For
```
```suggestion
Correspondingly, we also offer such functionality for ``function_pass``. For
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+
+
+Alternatively, users can also directly register a pass without using the
+decroators and then invoke it. Let's use ``Sequential`` to demo this scenario.
```
```suggestion
decorators and then invoke it. Let's use ``Sequential`` to demo this scenario.
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+
+The ``transform`` function here adds an ``abs`` function to the input module,
+but it could be any customized optimizations at the module level. After
+creating this `module_pass`, users can apply it on any Relay module. For
```
```suggestion
creating this ``module_pass``, users can apply it on any Relay module. For
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+
+This example not only illustrates how users can directly create a sequential
+pass using Python APIs (this could be applied to module- and function- level
+passes as well), but only explains we can build an optimization pipeline using
```
```suggestion
passes as well), but also explains how we can build an optimization pipeline using
```

weberlo commented on docs/dev/relay_pass_manager.rst:
```diff
@@ -0,0 +1,606 @@
+    # Create a module to perform optimizations.
+    mod = relay.Module({"main": func})
+    
+    # User can disable any passes that their don't want to execute by providing
```
```suggestion
    # User can disable any passes that they don't want to execute by providing
```

weberlo commented:
Awesome work!  This tutorial seems very comprehensive and has already helped me understand the pass manager a lot more.  I'll do another round of review at some point.

merrymercy commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,615 @@
+      // Passes need to be executed.
+      Array<Pass> passes;
+      // Passes are disabled during optimization
+      std::vector<std::string> disabled;
```
Why do not you use Array<StringImm> ?

zhiics replied:
Thanks for pointing out. `disabled` shouldn't belong here. We only have it in `PassContext` which is a `Array<tvm::Expr>`

zhiics commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,615 @@
+``pass_info`` is identical to what we just described in the module pass.
+``pass_func`` takes a function for optimization, it also needs a module as we
+may use it for reporting errors. A function could be annotated with
+"SkipOptimization" so that it will be ignored during optimization.
```
Yeah, I think `SkipOptimization` may be a better name. We should probably make the function name `SkipOptimization` as well.

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+Relay features a series of optimization passes which improve performance metrics
+of models such as mean inference, memory footprint, or power consumption for
+specific devices. There are a suite of standard optimizations as well as machine
+learning specific optimizations including constant folding, dead code
```
```suggestion
learning-specific optimizations including constant folding, dead code
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+specific devices. There are a suite of standard optimizations as well as machine
+learning specific optimizations including constant folding, dead code
+elimination, operator layout alteration, and operator fusion, etc. Each of these
+passes are structured as Relay-to-Relay transformations on the abstract syntax
```
```suggestion
passes is structured as a Relay-to-Relay transformation on the abstract syntax
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+passes are structured as Relay-to-Relay transformations on the abstract syntax
+tree (AST) using the analysis result collected during and/or before traversal.
+
+On one hand, many existing production compilers, such as GCC and LLVM, employ
```
It is confusing to have two paragraphs start with "On one hand" and "on the other" without explicitly setting up the contrast/tension beforehand. That is, there should be a sentence or paragraph first that sets up the tension between these two approaches.

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+handcrafted pass order.
+
+On the other hand, modern deep learning frameworks, such as Pytorch and MXNet
+Gluon, also have the tendency to enable pass style layer construction
```
```suggestion
Gluon, also have the tendency to enable pass-style layer construction
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+
+The design of the Relay pass infra is largely inspired by the the hierarchical
+pass manager used in LLVM and the block style containers used in the popular
+deep learning frameworks. The major goals of the pass infra include:
```
Are the two paragraphs before this one meant to set up a contrast between the two approaches? If so, this should explain how Relay's pass manager borrows from both.

zhiics replied:
Thank you for pointing out. I am actually not trying to setup a contrast. Instead, I want to say that we considered the advantages of both other approaches. I will rephrase.

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+#) enabling better programmatic orchestration of optimizations. This allows
+   users to flexibly customize and build their own optimization pipelines.
+
+#) providing a user friendly way to debug optimization passes.
```
```suggestion
#) providing a user-friendly way to debug optimization passes.
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+~~~~~~~~~~~
+
+We provide a ``PassInfo`` object to contain the basic information needed by
+a pass. ``name`` is the pass name, ``opt_level`` indicates at which optimization
```
Could there be a brief explanation of the opt_level mechanism? Or is it something entirely handled by the pass manager?

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+required/disabled passes, etc. For instance, we may have a configuration which
+performs all passes at ``opt_level=3`` with some disabled passes using
+``disabled_pass=xx`` provided by ``PassContext``. Now  we could glob all passes
+at ``opt_level=3`` and exclude those in disabled pass list.
```
```suggestion
at ``opt_level=3`` and exclude those in the disabled pass list.
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+``disabled_pass=xx`` provided by ``PassContext``. Now  we could glob all passes
+at ``opt_level=3`` and exclude those in disabled pass list.
+
+This class is designed as a thread local so that users can conveniently use the
```
What does "thread-local" mean in this context?

zhiics replied:
I will rephrase this. I wanted to say that we used a thread-local store to keep the pass context objects. Users can get the current context in a thread-safe way.

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+
+This class is designed as a thread local so that users can conveniently use the
+Python ``with`` syntax to perform optimizations under a certain configuration.
+Examples will be provided later to show how we can use both C++ and Python APIs
```
```suggestion
Examples will be provided later to show how we can use both the C++ and Python APIs
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+^^^^^^^^^^^^^^^
+
+The pass infra is designed in a hierarchical manner, and it could work at
+various granularity of a Relay program. A pure virtual class ``PassNode`` is
```
```suggestion
different granularities of Relay programs. A pure virtual class ``PassNode`` is
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+various granularity of a Relay program. A pure virtual class ``PassNode`` is
+introduced to serve as the base of the different optimization passes. This class
+contains several virtual methods that will be implemented by the
+subclasses at the module level, function level, or a sequence of passes..
```
```suggestion
subclasses at the level of modules, functions, or sequences of passes.
```
(grammatical parallelism)

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+The pass infra is designed in a hierarchical manner, and it could work at
+various granularity of a Relay program. A pure virtual class ``PassNode`` is
+introduced to serve as the base of the different optimization passes. This class
+contains several virtual methods that will be implemented by the
```
```suggestion
contains several virtual methods that must be implemented by the
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+                                const PassContext& pass_ctx) const = 0;
+    };
+
+The functor shows how a pass will be realized, i.e. it always works on a `Relay
```
```suggestion
The functor shows how a pass must be realized, i.e., it always works on a `Relay
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+    };
+
+The functor shows how a pass will be realized, i.e. it always works on a `Relay
+module`_ under a certain context. All passes are designed in a ``Module`` to
```
```suggestion
module`_ under a certain context. All passes are designed in a ``Module``-to-``Module``
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+
+The functor shows how a pass will be realized, i.e. it always works on a `Relay
+module`_ under a certain context. All passes are designed in a ``Module`` to
+``Module`` manner. Therefore, optimizations governed by the pass infra will
```
```suggestion
manner. Therefore, optimizations governed by the pass infra will
```
(combined with the previous suggestion)

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+``Module`` manner. Therefore, optimizations governed by the pass infra will
+always update the whole module.
+
+Several subclasses are created to implement different types of optimization
```
```suggestion
Several subclasses have been created to implement different types of optimization
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+always update the whole module.
+
+Several subclasses are created to implement different types of optimization
+passes, e.g. function-level passes, module-level passes, and sequential passes.
```
```suggestion
passes, e.g., function-level passes, module-level passes, and sequential passes.
```
Also it should be explained what "sequential passes" means somewhere. (I presume it means a sequence of passes; is that right?)

zhiics replied:
yes, it is right.

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+passes, e.g. function-level passes, module-level passes, and sequential passes.
+Each subclass itself could act as a pass manager. For instance, they could glob
+the require passes and execute them or build a dependency graph based on the
+given meta data. The full definition of them could be found in
```
```suggestion
given metadata. The full definition of them could be found in
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+Several subclasses are created to implement different types of optimization
+passes, e.g. function-level passes, module-level passes, and sequential passes.
+Each subclass itself could act as a pass manager. For instance, they could glob
+the require passes and execute them or build a dependency graph based on the
```
```suggestion
the required passes and execute them or build a dependency graph based on the
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+configure the compilation options, including optimization level and
+required/disabled passes, etc. For instance, we may have a configuration which
+performs all passes at ``opt_level=3`` with some disabled passes using
+``disabled_pass=xx`` provided by ``PassContext``. Now  we could glob all passes
```
Is "glob" a technical term used in this context? It seems a bit informal for this context otherwise.

weberlo replied:
I think it stems from the term "globbing" in Bash, but it probably _is_ too informal for the docs.  Maybe "collect" instead?

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+^^^^^^^^^^^^^^^^^^^
+
+Module level passes are geared mainly for global and inter-procedural
+optimizations (IPO), which is similar to the module pass used in LLVM. Some
```
```suggestion
optimizations (IPO), which are similar to the module pass used in LLVM. Some
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+
+Module level passes are geared mainly for global and inter-procedural
+optimizations (IPO), which is similar to the module pass used in LLVM. Some
+typical passes in Relay that need the global picture of a module, such as to
```
```suggestion
typical passes in Relay that need the global picture of a module, such as
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+Module level passes are geared mainly for global and inter-procedural
+optimizations (IPO), which is similar to the module pass used in LLVM. Some
+typical passes in Relay that need the global picture of a module, such as to
+a normal form conversion and lambda lifting, etc, fall in this set. At this
```
```suggestion
A-normal form conversion and lambda lifting, etc., fall into this set. At this
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+Sequential Passes
+^^^^^^^^^^^^^^^^^
+
+SequentialPass is similar to Pytorch nn.Sequential that contains a host of
```
We probably want `code::` annotations for terms like `SequentialPass` or `nn.Sequential` (just to be consistent)

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This example not only illustrates how users can directly create a sequential
+pass using Python APIs (this could be applied to module- and function- level
```
```suggestion
pass using Python APIs (this could be applied to module- and function-level
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,616 @@
+
+Only some simple APIs are needed for the frontend side. For example, we can
+provide users the following APIs to create and execute a pass (full
+implementation is provided in `python/tvm/relay/transform.py`_). The backend
```
I personally think "back-end" and "front-end" should be hyphenated but I don't know if everyone shares this opinion

weberlo replied:
I think they're common enough words now that we can drop the hyphen.

zhiics replied:
I will probably just keep them because it looks that they are also commonly used, including academic papers.

zhiics commented:
@anijain2305 @weberlo @shoubhik @slyubomirsky @merrymercy  Thanks for the reviews. I've addressed your comments. Please take another look when you have time.

weberlo commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,641 @@
+``Pass`` is the base class of all pass objects. All methods here are just simple
+wrappers that were implemented in the backend. They are defined for users to
+conveniently interact with the base class in Python. Only a functor is defined
+to invoke the execution of a pass.
```
This sentence is a little vague.  What functor?

zhiics replied:
Rephrase to ``__call__``

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,642 @@
+
+Relay features a series of optimization passes which improve performance metrics
+of models such as mean inference, memory footprint, or power consumption for
+specific devices. There are a suite of standard optimizations as well as machine
```
```suggestion
specific devices. There is a suite of standard optimizations as well as machine
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,642 @@
+specific devices. There are a suite of standard optimizations as well as machine
+learning-specific optimizations including constant folding, dead code
+elimination, operator layout alteration, and operator fusion, etc. Each of these
+passes are structured as a Relay-to-Relay transformation on the abstract syntax
```
```suggestion
passes is structured as a Relay-to-Relay transformation on the abstract syntax
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,642 @@
+passes are structured as a Relay-to-Relay transformation on the abstract syntax
+tree (AST) using the analysis result collected during and/or before traversal.
+
+However, as Relay evolves quickly, the need of a more systematic and efficient
```
```suggestion
However, as Relay evolves quickly, the need for a more systematic and efficient
```

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,642 @@
+tree (AST) using the analysis result collected during and/or before traversal.
+
+However, as Relay evolves quickly, the need of a more systematic and efficient
+way to manage these passes becomes imperative. This doc describes the design of
```
```suggestion
way to manage these passes is becoming apparent. This doc describes the design of
```

slyubomirsky replied:
I changed "imperative" to "apparent" because needs are inherently imperative

slyubomirsky commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,642 @@
+We focus on ease of extension for users, making it possible for users to quickly
+add new passes without loss of backward compatibility. The design contains both
+the backend and the frontend. The former implements the main logic of the pass
+infra. The latter provides simple APIs for users to interact with, i.e.
```
```suggestion
infra. The latter provides simple APIs for users to interact with, i.e.,
```

zhiics commented:
@slyubomirsky Thanks a lot.

@weberlo @shoubhik @anijain2305 Thank you all for the review. Please let me if you have any other comments or concerns.

weberlo commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,642 @@
+level the pass will be enabled, and ``required`` represents the passes that are
+required to execute a certain pass (see `include/tvm/relay/transform.h`_ for
+more details). For example, during registration of a pass (will be covered in
+later), the pass developers can specify the name of the pass, the optimiztion
```
```suggestion
later), the pass developers can specify the name of the pass, the optimization
```

weberlo commented on docs/dev/relay_pass_infra.rst:
```diff
@@ -0,0 +1,642 @@
+inspecting if it is specified as a required pass by the user. If it is still
+undetermined whether this pass is enabled, its ``opt_level`` will be checked.
+This pass will be enabled and therefore executed only when its optimization
+level not is less than the configured optimization level in the pass context.
```
```suggestion
level is not less than the configured optimization level in the pass context.
```

yzhliu commented:
Thanks everyone, this is now merged.
